player/lua/select.lua

```diff
@@ -353,6 +353,99 @@ mp.add_key_binding(nil, "select-audio-device", function ()
     })
 end)
 
+local function get_history_entry(line, seen, history_file_path)
+    local entry = utils.parse_json(line)
+
+    if not entry then
+        mp.msg.warn(line .. " in " .. history_file_path .. " is not valid JSON.")
+        return
+    end
+
+    local time, path, title = unpack(entry)
+
+    if not path then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    if seen[path] then
+        return
+    end
+
+    seen[path] = true
+
+    local status, date = pcall(os.date, "(%Y-%m-%d %H:%M) ", time)
+
+    if not status then
+        mp.msg.warn(line .. " in " .. history_file_path .. " has invalid data.")
+        return
+    end
+
+    return {
+        ["date"] = date,
+        ["path"] = path,
+        ["title"] = title,
+    }
+end
+
+mp.add_key_binding(nil, "select-watch-history", function ()
+    local history_file_path = mp.command_native({"expand-path",
+                                                 mp.get_property("watch-history-path")})
+    local history_file, error_message = io.open(history_file_path)
+    if not history_file then
+        show_warning(mp.get_property_native("save-watch-history")
+                     and error_message
+                     or "Enable --save-watch-history")
+        return
+    end
+
+    local lines = {}
+    for line in history_file:lines() do
+        table.insert(lines, line)
+    end
+    history_file:close()
+
+    local entries = {}
+    local seen = {}
+    for i = #lines, 1, -1 do
+        local entry = get_history_entry(lines[i], seen, history_file_path)
+        if entry then
+            entries[#entries + 1] = entry
+        end
+    end
+
+    local items = {}
+    local osd_playlist_entry = mp.get_property("osd-playlist-entry")
+    for i, entry in pairs(entries) do
+        local item = entry.path
+        if entry.title and osd_playlist_entry == "title" then
+            item = entry.title
+        elseif entry.title and osd_playlist_entry == "both" then
+            item = entry.title .. " (" .. entry.path .. ")"
+        end
+        items[i] = entry.date .. item
+    end
```

Comment:
I would put this back it the main loop, can be inside `if entry then` branch. I will see about making it more generic in another PR.